libsql-sqlite3/src/vectorfloat64.c

```diff
@@ -0,0 +1,240 @@
+**************************************************************************/
+
+void vectorF64Dump(const Vector *pVec){
+  double *elems = pVec->data;
+  unsigned i;
+  for(i = 0; i < pVec->dims; i++){
+    printf("%lf ", elems[i]);
+  }
+  printf("\n");
+}
+
+/**************************************************************************
+** Utility routines for vector serialization and deserialization
+**************************************************************************/
+
+static inline unsigned formatF64(double value, char *pBuf, int nBufSize){
+  sqlite3_snprintf(nBufSize, pBuf, "%g", value);
+  return strlen(pBuf);
+}
+
+static inline unsigned serializeF64(unsigned char *pBuf, double value){
+  u64 *p = (u64 *)&value;
+  pBuf[0] = *p & 0xFF;
+  pBuf[1] = (*p >> 8) & 0xFF;
+  pBuf[2] = (*p >> 16) & 0xFF;
+  pBuf[3] = (*p >> 24) & 0xFF;
+  pBuf[4] = (*p >> 32) & 0xFF;
+  pBuf[5] = (*p >> 40) & 0xFF;
+  pBuf[6] = (*p >> 48) & 0xFF;
+  pBuf[7] = (*p >> 56) & 0xFF;
+  return sizeof(double);
+}
+
+static inline double deserializeF64(const unsigned char *pBuf){
+  u64 value = 0;
+  value |= (u64)pBuf[0];
+  value |= (u64)pBuf[1] << 8;
+  value |= (u64)pBuf[2] << 16;
+  value |= (u64)pBuf[3] << 24;
+  value |= (u64)pBuf[4] << 32;
+  value |= (u64)pBuf[5] << 40;
+  value |= (u64)pBuf[6] << 48;
+  value |= (u64)pBuf[7] << 56;
+  return *(double *)&value;
+}
+
+size_t vectorF64SerializeToBlob(
+  const Vector *pVector,
+  unsigned char *pBlob,
+  size_t nBlobSize
+){
+  double *elems = pVector->data;
+  unsigned char *pPtr = pBlob;
+  size_t len = 0;
+  unsigned i;
+
+  assert( pVector->type == VECTOR_TYPE_FLOAT64 );
+  assert( pVector->dims <= MAX_VECTOR_SZ );
+  assert( nBlobSize >= pVector->dims * sizeof(double) );
+
+  for (i = 0; i < pVector->dims; i++) {
+    pPtr += serializeF64(pPtr, elems[i]);
+    len += sizeof(double);
+  }
+  return len;
+}
+
+size_t vectorF64DeserializeFromBlob(
+  Vector *pVector,
+  const unsigned char *pBlob,
+  size_t nBlobSize
+){
+  double *elems = pVector->data;
+  unsigned i;
+  pVector->type = VECTOR_TYPE_FLOAT64;
+  pVector->dims = nBlobSize / sizeof(double);
+
+  assert( pVector->dims <= MAX_VECTOR_SZ );
+  assert( nBlobSize >= vectorDataSize(pVector->type, pVector->dims) );
+
+  for(i = 0; i < pVector->dims; i++){
+    elems[i] = deserializeF64(pBlob);
+    pBlob += sizeof(double);
+  }
+  return vectorDataSize(pVector->type, pVector->dims);
+}
+
+void vectorF64Serialize(
+  sqlite3_context *context,
+  const Vector *pVector
+){
+  double *elems = pVector->data;
+  unsigned char *pBlob;
+  size_t nBlobSize;
+
+  assert( pVector->type == VECTOR_TYPE_FLOAT64 );
+  assert( pVector->dims <= MAX_VECTOR_SZ );
+
+  // allocate one extra trailing byte with vector blob type metadata
+  nBlobSize = vectorDataSize(pVector->type, pVector->dims) + 1;
+  pBlob = sqlite3_malloc64(nBlobSize);
+
+  if( pBlob ){
+    vectorF64SerializeToBlob(pVector, pBlob, nBlobSize - 1);
+    pBlob[nBlobSize - 1] = VECTOR_TYPE_FLOAT64;
```

Comment:
Why do we add this in vectorF64Serialize but we don't do this in vectorF32Serialize?

Reply:
We don't seem to use it in vectorF64DeserializeFromBlob

Reply:
We can assert last byte in `vectorF64DeserializeFromBlob`, yes, but the resolution of type happening in `vector.c` and `vectorfloatxx.c` source code just dealing with what it got.

In `vectorF32Serialize` I decided to not to fill trailing byte as even sized blobs are f32 by default and this way we will produce more "natural to human" blobs for F32 vectors.

Reply:
Ok I see it's used in  https://github.com/tursodatabase/libsql/pull/1531/files#diff-98fd1b40c774432f5288bd6ae7000632bbffb8687f15734977cb8cc794ed1a41R237